iot_packet_verifier/src/balances.rs

```diff
@@ -0,0 +1,130 @@
+use crate::{burner::Burn, pdas};
+use anchor_lang::AccountDeserialize;
+use data_credits::DelegatedDataCreditsV0;
+use futures_util::StreamExt;
+use helium_crypto::PublicKey;
+use solana_client::{client_error::ClientError, nonblocking::rpc_client::RpcClient};
+use solana_sdk::program_pack::Pack;
+use solana_sdk::pubkey::Pubkey;
+use sqlx::{Pool, Postgres};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+pub struct Balances {
+    pub provider: Arc<RpcClient>,
+    pub balances: Arc<Mutex<HashMap<PublicKey, Balance>>>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DebitError {
+    #[error("Sql error: {0}")]
+    SqlError(#[from] sqlx::Error),
+    #[error("Solana rpc error: {0}")]
+    RpcClientError(#[from] ClientError),
+    #[error("Anchor error: {0}")]
+    AnchorError(#[from] anchor_lang::error::Error),
+    #[error("Solana program error: {0}")]
+    ProgramError(#[from] solana_sdk::program_error::ProgramError),
+}
+
+impl Balances {
+    /// Fetch all of the current balances that have been actively burned so that
+    /// we have an accurate cache.
+    pub async fn new(
+        pool: &Pool<Postgres>,
+        sub_dao: &Pubkey,
+        provider: Arc<RpcClient>,
+    ) -> Result<Self, DebitError> {
+        let mut burns = sqlx::query_as("SELECT * FROM pending_burns").fetch(pool);
+
+        let mut balances = HashMap::new();
+
+        while let Some(Burn {
+            payer,
+            amount: burn_amount,
+            ..
+        }) = burns.next().await.transpose()?
+        {
+            // Look up the current balance of the payer
+            let payer = PublicKey::try_from(payer).unwrap();
```

Comment:
I am closing my eyes to multiple unwraps in here...see no evil
Also noting on the PublicKey vs PublicKeyBinary again